cfgrammar/src/lib/yacc/info.rs

```diff
@@ -0,0 +1,61 @@
+use crate::yacc::{YaccGrammarError, YaccGrammarWarning};
+
+/// Controls the collection of and collects information about warnings and other information
+/// that result during the construction of a `YaccGrammar`.  If `warnings_as_errors` is `true`
+/// does not collect any warnings.
+pub struct YaccGrammarInfo {
+    pub(crate) warnings_as_errors: bool,
+    warnings: Vec<YaccGrammarWarning>,
+}
+
+#[derive(Default)]
+pub struct YaccGrammarInfoBuilder {
+    warnings_as_errors: bool,
+}
+
+impl YaccGrammarInfoBuilder {
+    pub fn warnings_as_errors(mut self) -> Self {
+        self.warnings_as_errors = true;
+        self
+    }
+
+    pub fn build(self) -> YaccGrammarInfo {
+        YaccGrammarInfo {
+            warnings_as_errors: self.warnings_as_errors,
+            warnings: Vec::new(),
+        }
+    }
+}
+
+impl YaccGrammarInfo {
+    /// Returns a `YaccGrammarInfoBuilder` for configuring the behavior of a `YaccGrammarInfo`
+    pub fn builder() -> YaccGrammarInfoBuilder {
+        Default::default()
+    }
+
+    /// Returns a `YaccGrammarInfo` settings that are enabled are selected with a disposition
+    /// towards caution. For example, the treatment of warnings as errors would be enabled.
+    ///
+    /// The exact set of flags enabled by this function are not specified and subject to change.
+    pub fn cautious() -> Self {
+        Self::builder().warnings_as_errors().build()
+    }
+
+    /// If `warnings_as_errors` is `true` converts the warning to an error and returns `Some`.
+    /// Otherwise adds the warning for later usage and returns `None`.
+    pub(crate) fn add_warning(&mut self, warning: YaccGrammarWarning) -> Option<YaccGrammarError> {
```

Comment:
We can do that, but it does mean we'll no longer produce an `Err(...)` branch from the callee.  It is probably fine though if the callee knows nothing about this and produces `Ok` though.

Reply:
I removed all this in [c47e5c4](https://github.com/softdevteam/grmtools/pull/345/commits/c47e5c4135be21a4db2d57f1af6488b4488d55b0), some error conversion had to be added to `new_with_storaget`.

At some point I was worried that we get into a situation where we report an error, and then it *also* exists in warnings.
But that actually can't happen because `YaccGrammarInfo` can't escape from `new_with_storaget`.

But `new_with_storaget` is kind of the underlying need for conversion internally, I had tried to keep compatibility with the existing API here.